engine/lib/phx/src/engine/event_bus.rs

```diff
@@ -0,0 +1,318 @@
+use std::{
+    borrow::BorrowMut,
+    cell::Ref,
+    collections::{hash_map::Entry, BinaryHeap, HashMap},
+    sync::atomic::{AtomicU32, Ordering},
+};
+use strum::IntoEnumIterator;
+use strum_macros::EnumIter;
+use tracing::trace;
+
+use internal::ConvertIntoString;
+
+#[luajit_ffi_gen::luajit_ffi]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, EnumIter)]
+pub enum UpdatePass {
+    // Before physics update
+    PreSim,
+    // Physics update
+    Sim,
+    // After physics update
+    PostSim,
+    // Before frame render
+    PreFrame,
+    // Frame render
+    Frame,
+    // After frame render
+    PostFrame,
+    // Frame interpolation
+    FrameInterpolation,
+    // Before input handling
+    PreInput,
+    // Input handling
+    Input,
+    // After input handling
+    PostInput,
+}
+
+impl Default for UpdatePass {
+    fn default() -> Self {
+        UpdatePass::PreFrame
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum EventPayload {
+    Text(String),
+    Number(i32),
+}
+
+#[derive(Debug, Clone)]
+pub struct Subscriber {
+    tunnel_id: u32,
+    entity_id: Option<u32>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Event {
+    pub name: String,
+    pub priority: u16,
+    pub update_pass: UpdatePass,
+    pub subscribers: Vec<Subscriber>,
+    // pub payloads: Vec<EventPayload>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct MessageRequest {
+    priority: u16,
+    name: String,
+    stay_alive: bool,
+    completed: bool,
+}
+
+impl Ord for MessageRequest {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        other
+            .priority
+            .cmp(&self.priority)
+            .then_with(|| self.name.cmp(&other.name))
+    }
+}
+
+impl PartialOrd for MessageRequest {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+pub struct EventBus {
+    events: HashMap<String, Event>,
+    update_pass_map: HashMap<UpdatePass, BinaryHeap<MessageRequest>>,
+    next_tunnel_id: AtomicU32,
+    max_priority_locked: bool,
+}
+
+impl EventBus {
+    pub fn new() -> Self {
+        Self {
+            next_tunnel_id: AtomicU32::new(0),
+            update_pass_map: HashMap::new(),
+            events: HashMap::new(),
+            max_priority_locked: false,
+        }
+    }
+
+    pub fn lock_max_priority(&mut self) {
+        self.max_priority_locked = true;
+    }
+
+    pub fn dispatch(&mut self, update_pass: UpdatePass, lua: &mut Ref<Lua>) {
+        //* Can we handle the lua state differently? */
+        //* Maybe also with future potential multiple lua states in mind */
+        let globals = lua.globals();
+        let event_tunnels: Table = globals.get("EventTunnels").expect("Unknown table");
+
+        if let Some(message_heap) = self.update_pass_map.get_mut(&update_pass) {
+            let mut message_requests: Vec<_> = message_heap.drain().collect();
+
+            for message_request in message_requests.iter_mut() {
```

Comment:
I usually use `&mut message_requests` in such a loops instead of `iter_mut()`. Sema relates to `iter()` - just use `&`.

Reply:
Fair enough, I stand corrected